Implement FreeRTOS on Linux OSAL port

ismilak opened the pull request:
Implementing FreeRTOS linux port for OSAL. 
- FreeRTOS is used as a submodule in the project with POSIX port.c
- Specialised FreeRTOS config
- Network programming interface has been ported for linux, FreeRTOS socket API is not used in this project
- OS objects ported for using FreeRTOS API (thread, semaphore, etc.)
- Trickle timer ported to FreeRTOS timer thread
- Add new OSAL APIs for support FreeRTOS: kernel start and sleep API
- Implemented makefiles for FreeRTOS target
- Refactor sample app to support FreeRTOS environment

ismilak committed:
Implement freertos.target

ismilak committed:
add FreeRTOS-Kernel.git submodule

ismilak committed:
Add FreeRTOSConfig.h

ismilak committed:
Create empty osal implementation

ismilak committed:
Fix 'freertos/osal_common.h'

ismilak committed:
Refactor task set cancel type

ismilak committed:
Fix freertos.makefile include path

ismilak committed:
Add OSAL_FREERTOS_LINUX def to freertos make target

ismilak committed:
Implement CsmpAgentLib_sample_util header and source files

ismilak committed:
Implement OSAL start kernel API

ismilak committed:
Add thread and kernel asserts to main

ismilak committed:
Add freertos headers to freertos target

ismilak committed:
Implement FreeRTOS osal_task_create

ismilak committed:
Implement osal_kernel_start

ismilak committed:
Set minimal thread stack size to pthread size in FreeRTOSConfig

ismilak committed:
Implement FreeRTOS osal_task_cancel

ismilak committed:
Implement FreeRTOS osal semaphore calls

ismilak committed:
Implement FreeRTOS allocators

ismilak committed:
Implement FreeRTOS posix socket for osal

ismilak committed:
Implement get and set date time FreRTOS osal

ismilak committed:
Implement trickle timer with FreeRTOS timers

ismilak committed:
Implement osal_sleep_ms

ismilak committed:
Check return value of thread creation

ismilak committed:
Add missing doxy comments to 'osal/linux/osal_common.c'

ismilak committed:
Add doxy comments to 'osal/freertos/osal_common.c'

ismilak committed:
Fix osal refactor issues

paduffy reviewed (approved):
All these changes look reasonable to me.

ismilak committed:
Update README.md with FreeRTOS target content

ismilak commented:
Please review the markdown file update too. I updated the md file with minimal OSAL multi-target and  FreeRTOS build process information.

manojnacsl commented on sample/CsmpAgentLib_sample_util.c:
```diff
@@ -0,0 +1,670 @@
+#include <stdio.h>
```
Any specific reason for moving these from sample/CsmpAgentLib_sample.c to this new file? 
Original thought was to keep everything related to the sample app in one CsmpAgentLib_sample.c file. This file would not only have funcs handling the CSMP service handling but also the GET/PUT implementations as well.
What are your thoughts on this?

ismilak replied:
Hi! I thought we should keep CsmpAgentLib_sample.c only for the thread functions, since there will be more due to further target (efr32). Without separating the things the sample app c file will include  more thousands lines.
An other reason is the abstraction, what will be necessary later for efr32 target to customise the TLV handlers. It has been implemented in internal repo already to abstract the IPv6 address TLV for example.

manojnacsl replied:
True that TLV GET/PUT functions will increase over time as we keep implementing new TLV support. If we are expecting more code to handle the sample app's infra, we could go with the current approach - so we have all TLV GET/PUT implementations in a separate file.

manojnacsl reviewed (commented):
Overall changes look good - esp. under osal/freertos, osal/linux, src/coap and makefiles. Need to understand refactoring the existing sample/CsmpAgentLib_sample.c

manojnacsl commented on README.md:
```diff
@@ -1,10 +1,13 @@
 ## Overview
```
Looks good.

ismilak committed:
Fix typo in README.md and generate README.pdf

ismilak commented:
> > Please review the markdown file update too. I updated the md file with minimal OSAL multi-target and FreeRTOS build process information. @manojnacsl what is the required/used tool to generate pdf from markdown to have the same format?
> 
> @ismilak - We have been updating the README.md file using the Github's in-built editor, available within Github portal. A separate Developer's guide is maintained (docs/CSMP Developer Tutorial - 0v11.pdf) which is regularly updated and exported to pdf for sharing with the community. There are external tools viz., md2pdf, pandoc, markdown-pdf and others which could be used for conversion.

I fixed a typo in the readme and generated the pdf version of readme.md. Developer guide source is not available in the repository to modify the documentation. How can we update the Developer guide? Should we create a new one for new targets?
Anyway I think we should update dev guide when efr32 target will be introduced. FreeRTOS on Linux is the same runtime environment, there isn't any complicated step to execute the FreeRTOS binary, so I think mentioning in the current dev guide is not necessary. What do you think?

paduffy commented:
I’ll review and update the developer guide once the Silabs PRs are settled.

Cheers

From: Manojna CSL ***@***.***>
Sent: Thursday, May 30, 2024 1:15 AM
To: CiscoDevNet/csmp-agent-lib ***@***.***>
Cc: Paul Duffy (paduffy) ***@***.***>; Assign ***@***.***>
Subject: Re: [CiscoDevNet/csmp-agent-lib] Implement FreeRTOS on Linux OSAL port (PR #17)


Please review the markdown file update too. I updated the md file with minimal OSAL multi-target and FreeRTOS build process information. @manojnacsl<https://github.com/manojnacsl> what is the required/used tool to generate pdf from markdown to have the same format?

@ismilak<https://github.com/ismilak> - We have been updating the README.md file using the Github's in-built editor, available within Github portal. A separate Developer's guide is maintained (docs/CSMP Developer Tutorial - 0v11.pdf) which is regularly updated and exported to pdf for sharing with the community.

—
Reply to this email directly, view it on GitHub<https://github.com/CiscoDevNet/csmp-agent-lib/pull/17#issuecomment-2138697109>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/BHXULF655UDWTYYHJEYK6QTZE2YVTAVCNFSM6AAAAABH3Y6FMWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDCMZYGY4TOMJQHE>.
You are receiving this because you were assigned.Message ID: ***@***.******@***.***>>

manojnacsl commented:
> > > Please review the markdown file update too. I updated the md file with minimal OSAL multi-target and FreeRTOS build process information. @manojnacsl what is the required/used tool to generate pdf from markdown to have the same format?
> > 
> > 
> > @ismilak - We have been updating the README.md file using the Github's in-built editor, available within Github portal. A separate Developer's guide is maintained (docs/CSMP Developer Tutorial - 0v11.pdf) which is regularly updated and exported to pdf for sharing with the community. There are external tools viz., md2pdf, pandoc, markdown-pdf and others which could be used for conversion.
> 
> I fixed a typo in the readme and generated the pdf version of readme.md. Developer guide source is not available in the repository to modify the documentation. How can we update the Developer guide? Should we create a new one for new targets? Anyway I think we should update dev guide when efr32 target will be introduced. FreeRTOS on Linux is the same runtime environment, there isn't any complicated step to execute the FreeRTOS binary, so I think mentioning in the current dev guide is not necessary. What do you think?

@ismilak - It would be good to have one Developer's guide for all targets. @paduffy has confirmed he will review and update the Dev guide once all the ongoing FreeRTOS changes are concluded/committed.

manojnacsl reviewed (approved):
Overall looks good. Approving.

silabs-istvanr reviewed (approved):
For this phase it looks good to me
